vsphere/host_data_store_system_helper.go

```diff
@@ -0,0 +1,147 @@
+package vsphere
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/vmware/govmomi"
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/vim25/methods"
+	"github.com/vmware/govmomi/vim25/types"
+)
+
+// hostDatastoreSystemFromHostSystemID locates a HostDatastoreSystem from a
+// specified HostSystem managed object ID.
+func hostDatastoreSystemFromHostSystemID(client *govmomi.Client, hsID string) (*object.HostDatastoreSystem, error) {
+	hs, err := hostSystemFromID(client, hsID)
+	if err != nil {
+		return nil, err
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), defaultAPITimeout)
+	defer cancel()
+	return hs.ConfigManager().DatastoreSystem(ctx)
+}
+
+// availableScsiDisk checks to make sure that a disk is available for use in a
+// VMFS datastore, and returns the ScsiDisk.
+func availableScsiDisk(dss *object.HostDatastoreSystem, name string) (*types.HostScsiDisk, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultAPITimeout)
+	defer cancel()
+	disks, err := dss.QueryAvailableDisksForVmfs(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("cannot query available disks: %s", err)
+	}
+
+	var disk *types.HostScsiDisk
+	for _, d := range disks {
+		if d.CanonicalName == name {
+			disk = &d
+			break
+		}
+	}
+	if disk == nil {
+		return nil, fmt.Errorf("%s does not seem to be a disk available for VMFS", name)
+	}
+	return disk, nil
+}
+
+// diskSpecForCreate checks to make sure that a disk is available to be used to
+// create a VMFS datastore, specifically in its entirety, and returns a
+// respective VmfsDatastoreCreateSpec.
+func diskSpecForCreate(dss *object.HostDatastoreSystem, name string) (*types.VmfsDatastoreCreateSpec, error) {
+	disk, err := availableScsiDisk(dss, name)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultAPITimeout)
+	defer cancel()
+	options, err := dss.QueryVmfsDatastoreCreateOptions(ctx, disk.DevicePath)
+	if err != nil {
+		return nil, fmt.Errorf("could not get disk creation options for %s: %s", name, err)
+	}
+	var option *types.VmfsDatastoreOption
+	for _, o := range options {
+		if _, ok := o.Info.(*types.VmfsDatastoreAllExtentOption); ok {
+			option = &o
+			break
+		}
+	}
+	if option == nil {
+		return nil, fmt.Errorf("cannot use entire disk on device %s for datastore", name)
```

Comment:
could we add some more information to this message and change the formatting argument to `%q`?